asserts/account_key_test.go

```diff
@@ -590,6 +590,175 @@ func (aks *accountKeySuite) TestPublicKeyIsValidAt(c *C) {
+		{aks.since.AddDate(0, 3, 0), true},
+		{aks.since.AddDate(0, -2, 0), false},
+		{aks.until, false},
+		{aks.until.AddDate(0, 3, 0), false},
+		{aks.until.AddDate(0, -2, 0), true},
+	}
+
+	for _, t := range tests {
+		c.Check(asserts.AccountKeyIsKeyValidAssumingCurTimeWithin(accKey, t.timePt, t.timePt), Equals, t.valid)
+	}
+}
+
+func (aks *accountKeySuite) TestPublicKeyIsValidAssumingCurTimeWithinNoUntilPunctual(c *C) {
+	// With since and until, i.e. signing account-key expires.
+	// Key is valid for time >= since.
+	encoded := "type: account-key\n" +
+		"authority-id: canonical\n" +
+		"account-id: acc-id1\n" +
+		"name: default\n" +
+		"public-key-sha3-384: " + aks.keyID + "\n" +
+		aks.sinceLine +
+		fmt.Sprintf("body-length: %v", len(aks.pubKeyBody)) + "\n" +
+		"sign-key-sha3-384: Jv8_JiHiIzJVcO9M55pPdqSDWUvuhfDIBJUS-3VW7F_idjix7Ffn5qMxB21ZQuij" + "\n\n" +
+		aks.pubKeyBody + "\n\n" +
+		"AXNpZw=="
+	a, err := asserts.Decode([]byte(encoded))
+	c.Assert(err, IsNil)
+
+	accKey := a.(*asserts.AccountKey)
+
+	later := aks.until
+	tests := []struct {
+		timePt time.Time
+		valid  bool
+	}{
+		{aks.since, true},
+		{aks.since.AddDate(0, 3, 0), true},
+		{aks.since.AddDate(0, -2, 0), false},
+		{later, true},
+		{later.AddDate(0, 3, 0), true},
+	}
+
+	for _, t := range tests {
+		c.Check(asserts.AccountKeyIsKeyValidAssumingCurTimeWithin(accKey, t.timePt, t.timePt), Equals, t.valid)
+	}
+}
+
+func (aks *accountKeySuite) TestPublicKeyIsValidAssumingCurTimeWithinWithUntilInterval(c *C) {
+	// With since and until, i.e. signing account-key expires.
+	// Key is valid over [since, until)
+	encoded := "type: account-key\n" +
+		"authority-id: canonical\n" +
+		"account-id: acc-id1\n" +
+		"name: default\n" +
+		"public-key-sha3-384: " + aks.keyID + "\n" +
+		aks.sinceLine +
+		aks.untilLine +
+		fmt.Sprintf("body-length: %v", len(aks.pubKeyBody)) + "\n" +
+		"sign-key-sha3-384: Jv8_JiHiIzJVcO9M55pPdqSDWUvuhfDIBJUS-3VW7F_idjix7Ffn5qMxB21ZQuij" + "\n\n" +
+		aks.pubKeyBody + "\n\n" +
+		"AXNpZw=="
+	a, err := asserts.Decode([]byte(encoded))
+	c.Assert(err, IsNil)
+
+	accKey := a.(*asserts.AccountKey)
+
+	z := time.Time{}
+
+	tests := []struct {
+		earliest time.Time
+		latest   time.Time
+		valid    bool
+	}{
+		{aks.since, aks.until, true},
+		{aks.since, aks.since.AddDate(0, 3, 0), true},
+		{aks.since.AddDate(0, 1, 0), aks.since.AddDate(0, 3, 0), true},
+		{aks.since.AddDate(0, 1, 0), aks.until, true},
+		{aks.until, aks.until.AddDate(0, 3, 0), false},
+		{aks.until.AddDate(0, 2, 0), aks.until.AddDate(0, 3, 0), false},
+		{aks.since.AddDate(0, -1, 0), aks.since, true},
+		{aks.since.AddDate(0, -1, 0), aks.since.AddDate(0, 1, 0), true},
+		{aks.since.AddDate(0, -2, 0), aks.since.AddDate(0, -2, 0), false},
+		{aks.until.AddDate(0, -1, 0), aks.until.AddDate(0, 1, 0), true},
+		{aks.since, z, true},
+		{aks.since.AddDate(0, 1, 0), z, true},
+		{aks.since.AddDate(0, -3, 0), z, true},
+		{aks.until, z, false},
+		{aks.until.AddDate(0, 1, 0), z, false},
```

Comment:
this is missing cases where we don't know what the lower bound is i.e. where `earliest` is `z`. These cases aren't practical since we should always have a lower bound, but this function may see usage outside of it's current usage and we should be sure that the function works well if we don't have a lower time bound and we do have an upper possible current time bound

Reply:
I can add them but for for practical purposes time.Time{}  is not special as an earliest time, this is different from latest where it needs to be special cased and actively ignored

Reply:
yes that's what I mean is that it's not a special case of implementation, but I think we should have coverage of these cases, even if you just add them as a followup

Reply:
added some